components/builder-depot/src/server.rs

```diff
@@ -517,40 +505,43 @@ fn upload_package(req: &mut Request) -> IronResult<Response> {
         }
     }
 
-    match depot.datastore.packages.find(&ident) {
-        Ok(_) |
-        Err(dbcache::Error::EntityNotFound) => {
-            if let Some(_) = depot.archive(&ident) {
-                return Ok(Response::with((status::Conflict)));
-            }
-        }
+    let tempfile = match NamedTempFile::new() {
+        Ok(file) => file,
         Err(e) => {
-            error!("upload_package:1, err={:?}", e);
+            error!("could not create temporary file, err={:?}", e);
             return Ok(Response::with(status::InternalServerError));
         }
-    }
-
-    let filename = depot.archive_path(&ident);
-    try!(write_file(&filename, &mut req.body));
-    let mut archive = PackageArchive::new(filename);
+    };
+    let mut archive = try!(write_archive(&tempfile, &mut req.body));
     debug!("Package Archive: {:#?}", archive);
 
     let target_from_artifact = match archive.target() {
         Ok(target) => target,
         Err(e) => {
             info!("Could not read the target for {:#?}: {:#?}", archive, e);
-            remove_file(&archive.path).expect("Unable to remove temp file");
             return Ok(Response::with(status::UnprocessableEntity));
         }
     };
 
-    if depot.config.supported_target != target_from_artifact {
+    if !depot.config.supported_targets.contains(&target_from_artifact) {
         debug!("Unsupported package platform or architecture {}.",
                target_from_artifact);
-        remove_file(&archive.path).expect("Unable to remove temp file");
         return Ok(Response::with(status::NotImplemented));
     };
 
+    match depot.datastore.packages.find(&ident) {
+        Ok(_) |
+        Err(dbcache::Error::EntityNotFound) => {
+            if let Some(_) = depot.archive(&ident, &target_from_artifact) {
```

Comment:
This is a total nit and not important, but since you don't care about the value inside the `Some`, you could also write this as:

`if depot.archive(&ident, &target_from_artifact).is_some()`